service/history/timerQueueProcessor.go

```diff
@@ -588,96 +587,107 @@ Update_History_Loop:
+					{
+						t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.ScheduleToCloseTimeoutCounter)
+						if msBuilder.AddActivityTaskTimedOutEvent(ai.ScheduleID, ai.StartedID, timeoutType, nil) == nil {
 							return errFailedToAddTimeoutEvent
 						}
-
 						updateHistory = true
-						scheduleNewDecision = !msBuilder.HasPendingDecisionTask()
 					}
-				}
 
-			case workflow.TimeoutType_HEARTBEAT:
-				{
-					t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.HeartbeatTimeoutCounter)
-					lastHeartbeat := ai.LastHeartBeatUpdatedTime.Add(
-						time.Duration(ai.HeartbeatTimeout) * time.Second)
+				case workflow.TimeoutType_START_TO_CLOSE:
+					{
+						t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.StartToCloseTimeoutCounter)
+						if ai.StartedID != emptyEventID {
+							if msBuilder.AddActivityTaskTimedOutEvent(ai.ScheduleID, ai.StartedID, timeoutType, nil) == nil {
+								return errFailedToAddTimeoutEvent
+							}
+							updateHistory = true
+						}
+					}
 
-					if timerTask.VisibilityTimestamp.After(lastHeartbeat) {
+				case workflow.TimeoutType_HEARTBEAT:
+					{
+						t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.HeartbeatTimeoutCounter)
 						t.logger.Debugf("Activity Heartbeat expired: %+v", *ai)
-						if msBuilder.AddActivityTaskTimedOutEvent(scheduleID, ai.StartedID, timeoutType, nil) == nil {
+
+						if msBuilder.AddActivityTaskTimedOutEvent(ai.ScheduleID, ai.StartedID, timeoutType, nil) == nil {
 							return errFailedToAddTimeoutEvent
 						}
-
 						updateHistory = true
-						scheduleNewDecision = !msBuilder.HasPendingDecisionTask()
-					} else {
-						// Re-Schedule next heartbeat.
-						hbTimeoutTask, err := context.tBuilder.AddHeartBeatActivityTimeout(ai)
-						if err != nil {
-							return err
-						}
-						if hbTimeoutTask != nil {
-							timerTasks = append(timerTasks, hbTimeoutTask)
-						}
 					}
-				}
 
-			case workflow.TimeoutType_SCHEDULE_TO_START:
-				{
-					t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.ScheduleToStartTimeoutCounter)
-					if ai.StartedID == emptyEventID {
-						if msBuilder.AddActivityTaskTimedOutEvent(scheduleID, ai.StartedID, timeoutType, nil) == nil {
-							return errFailedToAddTimeoutEvent
+				case workflow.TimeoutType_SCHEDULE_TO_START:
+					{
+						t.metricsClient.IncCounter(metrics.TimerTaskActivityTimeoutScope, metrics.ScheduleToStartTimeoutCounter)
+						if ai.StartedID == emptyEventID {
+							if msBuilder.AddActivityTaskTimedOutEvent(ai.ScheduleID, ai.StartedID, timeoutType, nil) == nil {
+								return errFailedToAddTimeoutEvent
+							}
+							updateHistory = true
 						}
-
-						updateHistory = true
-						scheduleNewDecision = !msBuilder.HasPendingDecisionTask()
 					}
 				}
+			} else {
+				// See if we have next timer in list to be created.
+				if !td.TaskCreated || td.ActivityID == scheduleID {
```

Comment:
Do you have test coverage for a case where scheduledToStartTimeout > startToCloseTimeout ? If not, can you add one ?